Mathlib/Topology/CWComplex.lean

```diff
@@ -0,0 +1,140 @@
+/-
+Copyright (c) 2024 Elliot Dean Young and Jiazhen Xia. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiazhen Xia, Elliot Dean Young
+-/
+
+import Mathlib.Topology.Category.TopCat.Limits.Basic
+import Mathlib.Analysis.InnerProductSpace.PiL2
+import Mathlib.CategoryTheory.Functor.OfSequence
+
+/-!
+# CW-complexes
+
+This file defines (relative) CW-complexes.
+
+## Main definitions
+
+* `RelativeCWComplex`: A relative CW-complex is the colimit of an expanding sequence of subspaces
+`sk i` (called the `(i-1)`-skeleton) for `i ≥ 0`, where `sk 0` (i.e., the `(-1)`-skeleton) is an
+arbitrary topological space, and each `sk (n+1)` (i.e., the `n`-skeleton) is obtained from `sk n`
+(i.e., the `(n-1)`-skeleton) by attaching `n`-disks.
+
+* `CWComplex`: A CW-complex is a relative CW-complex whose `sk 0` (i.e., `(-1)`-skeleton) is empty.
+
+## References
+
+The definition of CW-complexes follows David Wärn's suggestion at
+https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Do.20we.20have.20CW.20complexes.3F/near/231769080
+-/
+
+open CategoryTheory
+
+universe u
+
+namespace RelativeCWComplex
+
+/-- The `n`-sphere is the set of points in ℝⁿ⁺¹ whose norm equals `1`,
+endowed with the subspace topology. -/
+noncomputable def sphere (n : ℤ) : TopCat.{u} :=
+  TopCat.of <| ULift <| Metric.sphere (0 : EuclideanSpace ℝ <| Fin <| Int.toNat <| n + 1) 1
+
+/-- The `n`-disk is the set of points in ℝⁿ whose norm is at most `1`,
+endowed with the subspace topology. -/
+noncomputable def disk (n : ℤ) : TopCat.{u} :=
+  TopCat.of <| ULift <| Metric.closedBall (0 : EuclideanSpace ℝ <| Fin <| Int.toNat n) 1
+
+/-- `𝕊 n` denotes the `n`-sphere. -/
+scoped notation "𝕊 "n => sphere n
+
+/-- `𝔻 n` denotes the `n`-disk. -/
+scoped notation "𝔻 "n => disk n
```

Comment:
I feel like this might be useful more generally than just CW-complexes. Should it move to a `TopCat` file?

Reply:
I agree. (This might be useful for homology of `𝕊 n`, degree of maps from `𝕊 n` to `𝕊 n`, etc.)

Could we move `sphere`, `disk`, and their notations to `Topology/Category/TopCat/Basic.lean`? In the future, if there are more constructions of objects in `TopCat`, we could then move them to a new file.

Reply:
Moving `sphere` and `disk` to `Topology/Category/TopCat/Basic.lean` leads to a conflict between the Lean-3-style notation `⟪x, y⟫_𝕜` for inner product and the (Lean 4) notation `x |_ₗ U ⟪e⟫` for `TopCat.Presheaf.restrict`, and hence a syntax error in `Topology/Sheaves/Presheaf.lean`

https://github.com/leanprover-community/mathlib4/blob/2eb00cae7edccf4e54238db2cb4b0a7dff59fbdd/Mathlib/Topology/Sheaves/Presheaf.lean#L125

with error message `error: ././././Mathlib/Topology/Sheaves/Presheaf.lean:125:12: unexpected token '⟫'; expected ','`.

The import links are shown below, where the red link is for the `EuclideanSpace` in the definition of `sphere` and `disk`.

![20240909_import_links_annotated](https://github.com/user-attachments/assets/6995287f-2e8e-413f-a2bb-e5b1ea196f9e)

I tried making the Lean-3-style notation `⟪x, y⟫_𝕜` local, but this leads to more errors and I'm reluctant to modify existing code unrelated to this PR.

(There is also a conflict in `Topology/Category/TopCat/Limits/Basic.lean` where the tactic `by continuity` presumably gets overwhelmed by newly imported lemmas. This [can be fixed](https://github.com/leanprover-community/mathlib4/pull/12502/commits/6ed65746447583d849c1af5cb71820d43284885d) by using the lemma `continuous_const` instead of the tactic.)

Could anyone help me decide the best way to resolve these conflicts? If Lean had a mechanism to `import` only a part of a file this would be easy, but I didn't find any instructions in the documentation.

Reply:
Oh, that looks bad. Could you start a discussion on Zulip and then tag this PR as `awaiting-zulip`?

Reply:
Sure. I've started [a Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Notation.20conflict.20in.20CW-complex.20PR).

Reply:
The notation conflict is resolved by PR #16672.

Another conflict is in `Topology/Category/TopCat/Limits/Basic.lean`

https://github.com/leanprover-community/mathlib4/blob/e992840b976c4ebb76851871c8cbb559374331af/Mathlib/Topology/Category/TopCat/Limits/Basic.lean#L187

where the tactic `by continuity` presumably gets overwhelmed by newly imported definitions, resulting in an `aesop` timeout. I didn't look into the exact cause but I was able to fix it simply by using the theorem `continuous_const`

https://github.com/leanprover-community/mathlib4/blob/e992840b976c4ebb76851871c8cbb559374331af/Mathlib/Topology/Basic.lean#L1482-L1484

instead of the tactic. Is it OK to do the fix directly in the current PR (commit https://github.com/leanprover-community/mathlib4/pull/12502/commits/c4a602a942a0936368f1ff354e4214eeae8624c7 )?

With these two conflicts resolved, I was finally able to move `sphere` and `disk` into `TopCat/Basic.lean` (commit https://github.com/leanprover-community/mathlib4/pull/12502/commits/8f1214a13e171d1e37a59287f13522cf38a38a36 ).